packages/connector-discord/src/index.ts

```diff
@@ -0,0 +1,152 @@
+  CreateConnector,
+  SocialConnector,
+  ConnectorType,
+} from '@logto/connector-core';
+import { assert, conditional } from '@silverhand/essentials';
+import got, { HTTPError } from 'got';
+
+import {
+  defaultMetadata,
+  scope,
+  authorizationEndpoint,
+  accessTokenEndpoint,
+  defaultTimeout,
+  userInfoEndpoint,
+} from './constant';
+import {
+  DiscordConfig,
+  discordConfigGuard,
+  authResponseGuard,
+  accessTokenResponseGuard,
+  userInfoResponseGuard,
+} from './types';
+
+const getAuthorizationUri =
+  (getConfig: GetConnectorConfig): GetAuthorizationUri =>
+  async ({ state, redirectUri }) => {
+    const config = await getConfig(defaultMetadata.id);
+    validateConfig<DiscordConfig>(config, discordConfigGuard);
+
+    const queryParameters = new URLSearchParams({
+      client_id: config.clientId,
+      redirect_uri: redirectUri,
+      response_type: 'code',
+      scope,
+      state,
+    });
+
+    return `${authorizationEndpoint}?${queryParameters.toString()}`;
+  };
+
+export const getAccessToken = async (
+  config: DiscordConfig,
+  codeObject: { code: string; redirectUri: string }
+) => {
+  const { code, redirectUri } = codeObject;
+
+  const { clientId: client_id, clientSecret: client_secret } = config;
+
+  const httpResponse = await got.post(accessTokenEndpoint, {
+    form: {
+      client_id,
+      client_secret,
+      grant_type: 'authorization_code',
+      code,
+      redirect_uri: redirectUri,
+    },
+    timeout: defaultTimeout,
+  });
+
+  const result = accessTokenResponseGuard.safeParse(JSON.parse(httpResponse.body));
+
+  if (!result.success) {
+    throw new ConnectorError(ConnectorErrorCodes.InvalidResponse, result.error.message);
+  }
+
+  const { access_token: accessToken } = result.data;
+
+  assert(accessToken, new ConnectorError(ConnectorErrorCodes.SocialAuthCodeInvalid));
+
+  return { accessToken };
+};
+
+/* eslint-disable complexity */
+const getUserInfo =
+  (getConfig: GetConnectorConfig): GetUserInfo =>
+  async (data) => {
+    const { code, redirectUri } = await authorizationCallbackHandler(data);
+    const config = await getConfig(defaultMetadata.id);
+    validateConfig<DiscordConfig>(config, discordConfigGuard);
+    const { accessToken } = await getAccessToken(config, { code, redirectUri });
+
+    try {
+      const httpResponse = await got.get(userInfoEndpoint, {
+        headers: {
+          authorization: `Bearer ${accessToken}`,
+        },
+        timeout: defaultTimeout,
+      });
+
+      const result = userInfoResponseGuard.safeParse(JSON.parse(httpResponse.body));
+
+      if (!result.success) {
+        throw new ConnectorError(ConnectorErrorCodes.InvalidResponse, result.error.message);
+      }
+
+      const { id, username: name, avatar, email, verified } = result.data;
+
+      return {
+        id,
+        name,
+        avatar: conditional(avatar && `https://cdn.discordapp.com/avatars/${id}/${avatar}`),
+        email: conditional(verified && email),
```

Comment:
I think maybe we should return both `email` and `email_verified (boolean)` in the response. @darcyYe  what's your take on this one?

Reply:
You are right, but I did not see code changes in the `users` schema for now. Let's align the timeline of changing the schema since we can not save the value of `verified` before we have that done. So does the rename of `avatar` to `picture`.

Reply:
Yes, let's do that later